frontend/src/metabase-lib/parameters/utils/click-behavior.ts

```diff
@@ -0,0 +1,330 @@
+import _ from "underscore";
+import { getIn } from "icepick";
+
+import type {
+  ClickBehavior,
+  Dashboard,
+  DashboardCard,
+  DatasetColumn,
+  DatetimeUnit,
+  Parameter,
+  ParameterType,
+  ParameterValueOrArray,
+  UserAttribute,
+} from "metabase-types/api";
+import { isImplicitActionClickBehavior } from "metabase-types/guards";
+import type { ValueAndColumnForColumnNameDate } from "metabase/lib/formatting/link";
+import { parseTimestamp } from "metabase/lib/time";
+import {
+  formatDateTimeForParameter,
+  formatDateToRangeForParameter,
+} from "metabase/lib/formatting/date";
+import {
+  dimensionFilterForParameter,
+  variableFilterForParameter,
+} from "metabase-lib/parameters/utils/filters";
+import type { Dimension } from "metabase-lib/types";
+import { isa, isDate } from "metabase-lib/types/utils/isa";
+import { TYPE } from "metabase-lib/types/constants";
+import TemplateTagVariable from "metabase-lib/variables/TemplateTagVariable";
+import { TemplateTagDimension } from "metabase-lib/Dimension";
+import type Question from "metabase-lib/Question";
+import type { ClickObjectDataRow } from "metabase-lib/queries/drills/types";
+
+interface SourceFilters {
+  column: (column: DatasetColumn) => boolean;
+  parameter: (parameter: Parameter) => boolean;
+  userAttribute: (userAttribute: string) => boolean;
+}
+
+interface ParameterTarget {
+  type: "parameter";
+  id: Parameter["id"];
+}
+
+interface Target {
+  id: Parameter["id"];
+  name: Parameter["name"];
+  target: ParameterTarget;
+  sourceFilters: SourceFilters;
+}
+
+export function getDataFromClicked({
+  extraData: { dashboard, parameterValuesBySlug = {}, userAttributes } = {},
+  dimensions = [],
+  data = [],
+}: {
+  extraData?: {
+    dashboard?: Dashboard;
+    parameterValuesBySlug?: Record<string, ParameterValueOrArray>;
+    userAttributes?: UserAttribute[] | null;
+  };
+  dimensions: Dimension[];
+  data: (ClickObjectDataRow & {
+    clickBehaviorValue: ClickObjectDataRow["value"];
+  })[];
+}): ValueAndColumnForColumnNameDate {
+  const column = [
+    ...dimensions,
+    ...data.map(d => ({
+      column: d.col,
+      // When the data is changed to a display value for use in tooltips, we can set clickBehaviorValue to the raw value for filtering.
+      value: d.clickBehaviorValue || d.value,
+    })),
+  ]
+    .filter(d => d.column != null)
+    .reduce<ValueAndColumnForColumnNameDate["column"]>(
+      (acc, { column, value }) => {
+        if (!column) {
+          return acc;
+        }
+
+        const name = column.name.toLowerCase();
+
+        if (acc[name] === undefined) {
+          return { ...acc, [name]: { value, column } };
+        }
+
+        return acc;
+      },
+      {},
+    );
+
+  const dashboardParameters = (dashboard?.parameters || []).filter(
+    ({ slug }) => parameterValuesBySlug[slug] != null,
+  );
+
+  const parameterByName =
+    dashboard == null
+      ? {}
+      : Object.fromEntries(
```

Comment:
Previously `parameterByName` and `parameter` were defined using underscore helpers but it produced incorrect types, so I had to change it.